### ISSUE TITLE
FUN.autodeps produce NoOp dependencies

### ISSUE THREAD

**Issue by mathias-lang-sociomantic** [sociomantic-tsunami/makd#76]
In one of our applications, I use `FUN.autodeps`.
Recently, as I was updating it on a server, I realized it failed to update `libdrizzle-redux`.
The old `libdrizzle-redux` used to compile the package was `v5.6.0`, but the server only had `v5.4.1`.
The newly deployed app required `v6.0.0` (hence a different soname, but at that time the package had the same name), and as a result failed to start because `apt-get` did not update the dependency.

I did a bit of investigation. First the output of `autodeps` is the following:
```
['libbz2-1.0', 'libc6', 'libdrizzle-redux', 'libgcc1', 'libgcrypt20', 'libgeoip1', 'libglib2.0-0', 'libgpg-error0', 'liblzo2-2', 'libmcrypt4', 'libpcre3', 'libssl1.0.0', 'libstdc++6', 'libtokyocabinet9', 'zlib1g']
```

When I run `dpkg -I` on the resulting package, I get:
```
[...]
 Depends: libbz2-1.0, libc6, libdrizzle-redux, libgcc1, libgcrypt20, libgeoip1, libglib2.0-0, libgpg-error0, liblzo2-2, libmcrypt4, libpcre3, libssl1.0.0, libstdc++6, libtokyocabinet9, zlib1g, libbz2-1.0, libc6, libdrizzle-redux, libgcc1, libgcrypt20, libgeoip1, libglib2.0-0, libgpg-error0, liblzo2-2, libmcrypt4, libpcre3, libssl1.0.0, libstdc++6, libtokyocabinet9, zlib1g, libbz2-1.0, libc6, libdrizzle-redux, libgcc1, libgcrypt20, libgeoip1, libglib2.0-0, libgpg-error0, liblzo2-2, libmcrypt4, libpcre3, libssl1.0.0, libstdc++6, libtokyocabinet9, zlib1g
[...]
```

As you can notice, the package includes 3 binaries, which have the same dependencies, so dependencies appear 3 times in the output.

When I run `apt-cache showpkg mypackage`, I get the following:
```
Reverse Depends: 
Dependencies: 
2.3.1-xenial - libc6 (0 (null)) libdrizzle-redux (0 (null)) libgcc1 (0 (null)) libgcrypt20 (0 (null)) libgeoip1 (0 (null)) libglib2.0-0 (0 (null)) libgpg-error0 (0 (null)) liblzo2-2 (0 (null)) libmcrypt4 (0 (null)) libpcre3 (0 (null)) libssl1.0.0 (0 (null)) libstdc++6 (0 (null)) zlib1g (0 (null)) libc6 (0 (null)) libdrizzle-redux (0 (null)) libgcc1 (0 (null)) libgcrypt20 (0 (null)) libgeoip1 (0 (null)) libglib2.0-0 (0 (null)) libgpg-error0 (0 (null)) liblzo2-2 (0 (null)) libmcrypt4 (0 (null)) libpcre3 (0 (null)) libssl1.0.0 (0 (null)) libstdc++6 (0 (null)) zlib1g (0 (null)) libc6 (0 (null)) libdrizzle-redux (0 (null)) libgcc1 (0 (null)) libgcrypt20 (0 (null)) libgeoip1 (0 (null)) libglib2.0-0 (0 (null)) libgpg-error0 (0 (null)) liblzo2-2 (0 (null)) libmcrypt4 (0 (null)) libpcre3 (0 (null)) libssl1.0.0 (0 (null)) libstdc++6 (0 (null)) zlib1g (0 (null)) 
```

The man page of `apt-cache showpkg` tell me to read the source code (!), so I did, and here's the [relevant section](https://github.com/Debian/apt/blob/2a90aa7a064047fb1c8783b31720cd345018ca4a/apt-pkg/pkgcache.h#L156). In short, `0` means the dependency is a no-op, and there's no associated version.

TL;DR: `FUN.autodeps` only ensures the package is installed based on the package name, but does not store version information, which can result in failures and manual dependency upgrade.

**Comment by mathias-lang-sociomantic** [comment]
After investigation, packages are indeed installed on the first install. However they are never upgraded and no version check is performed, which can lead to surprising effects.

**Comment by leandro-lucarella-sociomantic** [comment]
> TL;DR: FUN.autodeps only provides suggestions which are not even output to the screen.

What do you mean that they are not output to the screen?

**Comment by leandro-lucarella-sociomantic** [comment]
> After investigation, packages are indeed installed on the first install. However they are never upgraded and no version check is performed, which can lead to surprising effects.

OK, so the problem seems to be that `autodeps` doesn't specify any version. What dpkg does for what `autodeps` generates seems to make sense: if you don't declare dependencies with versions, it is expected that any package version will satisfy the dependency.

Maybe `autodeps` could specify versions as `pkg (>= {installed version})`, but this could be a bit restrictive. I'm not really sure about what to do here, to handle dependencies properly one should declare them explicitly. OTOH, `autodeps` was inspired by the deb helper [dh_shlibdeps(1)](http://manpages.ubuntu.com/manpages/xenial/en/man1/dh_shlibdeps.1.html) (I don't remember why I didn't use that directly), so it might be interesting to know what that helper does exactly.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, the real program calculating dependencies is [dpkg-shlibdeps](http://manpages.ubuntu.com/manpages/xenial/en/man1/dpkg-shlibdeps.1.html) and I don't know what or how it does it, but it looks for minimal dependencies. I didn't use it because it needs to work over a `debian/control` file, but I just quickly hacked one to try it out and for this example I got:

```
/tmp ✔ $ cat  debian/control 
Source: test
Standards-Version: 3.8.4
Section: libs
Description: Test

Package: test
Depends: ${shlibs:dependency-field}
Architecture: amd64
Description: Test
luca@labs-064:/tmp ✔ $ dpkg-shlibdeps /usr/bin/vim 
dpkg-shlibdeps: warning: binaries to analyze should already be installed in their package's directory

/tmp ✔ $ cat debian/substvars 
shlibs:Depends=libacl1 (>= 2.2.51-8), libbonoboui2-0 (>= 2.15.1), libc6 (>= 2.15), libgdk-pixbuf2.0-0 (>= 2.25.2), libglib2.0-0 (>= 2.31.18), libgnome-2-0, libgnomeui-0 (>= 2.22.0), libgpm2 (>= 1.20.4), libgtk2.0-0 (>= 2.24.0), libice6 (>= 1:1.0.0), liblua5.2-0, libpango-1.0-0 (>= 1.32.4), libperl5.22 (>= 5.22.1), libpython3.5 (>= 3.5.0~b1), libruby2.3 (>= 2.3.0~preview2), libselinux1 (>= 1.32), libsm6, libtcl8.6 (>= 8.6.0), libtinfo5 (>= 6), libx11-6, libxt6

/tmp ✔ $ dpkg -l libacl1  libbonoboui2-0 libc6 libgdk-pixbuf2.0-0 libglib2.0-0 libgnomeui-0 libgpm2
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                          Version             Architecture        Description
+++-=============================-===================-===================-===============================================================
ii  libacl1:amd64                 2.2.52-3            amd64               Access control list shared library
ii  libbonoboui2-0:amd64          2.24.5-3ubuntu1     amd64               Bonobo UI library - runtime
ii  libc6:amd64                   2.23-0ubuntu7       amd64               GNU C Library: Shared libraries
ii  libc6:i386                    2.23-0ubuntu7       i386                GNU C Library: Shared libraries
ii  libgdk-pixbuf2.0-0:amd64      2.32.2-1ubuntu1.2   amd64               GDK Pixbuf library
ii  libglib2.0-0:amd64            2.48.2-0ubuntu1     amd64               GLib library of C routines
ii  libglib2.0-0:i386             2.48.2-0ubuntu1     i386                GLib library of C routines
ii  libgnomeui-0:amd64            2.24.5-3.1          amd64               GNOME user interface library - runtime files
ii  libgpm2:amd64                 1.20.4-6.1          amd64               General Purpose Mouse - shared library
```

It doesn't even pick a version for `libgnome-2-0` for example, and it picks an older version for `libacl1`.

The *good news* are the [script source code](https://anonscm.debian.org/cgit/dpkg/dpkg.git/tree/scripts/dpkg-shlibdeps.pl) is not that big, so we could do an utility based on it that doesn't require a `debian/control` file just using the same libraries. The bad news is it's Perl.

**Comment by leandro-lucarella-sociomantic** [comment]
As a first approximation maybe just specifying the dependencies with `pkg (>= {current})` might be a better default than what we have now. But again, these tools are just to simplify adding the basic dependencies, like `libc` and such. If your application directly and deeply uses some library, you should specify that dependency manually. (the problem now is we don't have a way to *override* the manually set dependency in `autodeps`, so we'll have it duplicated, and I'm having the feeling we'll end up reimplementing `dpkg` if we follow this route :S).

**Comment by leandro-lucarella-sociomantic** [comment]
I'm looking into fixing this by using `dpkg-shlibdeps`, but still the de-duplication of dependencies is quite complex. I definitely don't want to parse every dependency exactly and start making decision on what to do with packages that appear more than once in the list but have a different version specification.

@mathias-lang-sociomantic was the repetition part of the problem or just an aesthetic issue?

**Comment by leandro-lucarella-sociomantic** [comment]
Avoiding duplicated dependencies when specifying multiple binaries (given that it is done in only one `autodeps()` call) should be fine, as I can pass all the binaries to `dpkg-shlibdeps` at once.

**Comment by leandro-lucarella-sociomantic** [comment]
Changing milestone as this is a bug, so it should probably go to v1.8.x

**Comment by leandro-lucarella-sociomantic** [comment]
Going back to v1.10.0 milestone, it seems like too much of a big change for a patch-release, it doesn't feel right risk-wise. #85 should fix this, mostly.

**Comment by mathias-lang-sociomantic** [comment]
> @mathias-lang-sociomantic was the repetition part of the problem or just an aesthetic issue?

I don't remember any issue. I was wondering how dkg would behave with different versions required for the same package but couldn't test anyway. I'll test the PR next week hopefully.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, soonish test is appreciated as I want to release v1.10.0 soon! :)

**Comment by llucax** [comment]
OK, testing soon is very welcome as I would like to release v1.10.0 soon too.
